torch/testing/_internal/common_methods_invocations.py

```diff
@@ -76,22 +76,27 @@ class SampleInput(object):
     """Represents sample inputs to a function."""
 
     # output_process_fn_grad is a function that modifies the output of op compatible with input
-    __slots__ = ['input', 'args', 'kwargs', 'output_process_fn_grad']
+    __slots__ = ['input', 'args', 'kwargs', 'output_process_fn_grad', 'broadcasts_self']
 
-    def __init__(self, input, *, args=tuple(), kwargs=None, output_process_fn_grad=None):
+    def __init__(self, input, *, args=tuple(), kwargs=None, output_process_fn_grad=None, broadcasts_self=False):
```

Comment:
This approach is reasonable but I still think we should filter sample_inputs() by whether the samples will be used for inplace computations or not. However, let me get a few other opinions on this. Maybe this is a better approach.

Reply:
Sorry I neglected to update this, @kshitij12345. General consensus is that it'd be preferable to filter sample_inputs() by what's inplace for now. We may want to revise that decision later, however.

Reply:
IIUC,
We will have to update the signature for every `sample_inputs_*` with 
```
sample_inputs_*(op_info, device, dtype, requires_grad, is_inplace_variant=False)
```
?

Reply:
We could do that (and make is_inplace_variant kwarg-only). That would be the most discoverable thing.

But we don't have to. The operations for testing the inplace variant can request sample inputs by passing the for_inplace_variant (we can work on the name) kwarg like this:

```
try:
  samples = op.sample_inputs(..., for_inplace_variant=True)
except TypeError as te:
  samples = op.sample_inputs(...) 
```

Then functions which support this option can implement it as:

```
def sample_inputs_foo(..., *, for_inplace_variant=False)
```

A nicer (but more disruptive) solution would be to make sample_inputs take **kwargs, and then the functions that want to use "for_inplace_variant" or similar options can query for it from the kwarg dict.

 What would you think of an approach like that?

Reply:
I like the second approach (though more disruptive). Will try that. 
Thanks!

Reply:
For `test_variant_consistency_eager`, 

https://github.com/pytorch/pytorch/blob/0d81528a470f4dfa57461d19d24ed9baba904caf/test/test_ops.py#L204-L242

We will have to put call for `sample_inputs` inside the variants loop. Since call to sample_input has to actually materialize all the tensors, this will result in some performance regression for this test. Let me know if that sounds acceptable.

Reply:
That's a great point, and that function could probably use a small refactoring. I agree with you but think we can mitigate the damage:

```
samples = op.sample_inputs(device, dtype, requires_grad=op.supports_autograd)

for sample in samples:
  # Acquires variants (method variant, inplace variant, aliases)
  method = op.get_method()
  inplace = op.get_inplace()

  # list of all inplace ops: inplace variant + alias inplace variants if exist
  inplace_ops = [inplace, ]

  aliases = []
  for a_op in op.aliases:
    aliases.append(a_op.op)
    aliases.append(a_op.method_variant)
    aliases.append(a_op.inplace_variant)
    inplace_ops.append(a_op.inplace_variant)
  
  aliases = tuple(aliases)
  inplace_ops = tuple(v for v in inplace_ops if v is not None)
  variants = (v for v in (method, inplace) + aliases if v is not None)
```

What's weird about this is that the test is rebuilding its understand of the operator's variants on each sample. This is unnecessary since variants are sample-independent. We should lift that section out of the for loop and acquire the variants once upfront.

Now let's look at the rest of the function:

```
# Computes function forward and backward values
sample.input.grad = None
expected_forward = op(sample.input, *sample.args, **sample.kwargs)
expected_grad = None

if (op.supports_autograd and isinstance(expected_forward, torch.Tensor)):
  expected_forward.sum().backward()
  expected_grad = sample.input.grad

  for variant in variants:
    sample.input.grad = None
    cloned = clone_input_helper(sample.input) if variant in inplace_ops else sample.input
    variant_forward = variant(cloned, *sample.args, **sample.kwargs)
    self.assertEqual(expected_forward, variant_forward)

    if expected_grad is not None and (variant not in inplace_ops or op.supports_inplace_autograd):
      variant_forward.sum().backward()
      self.assertEqual(expected_grad, sample.input.grad)
```

I think we can get rid of the `for variant in variants` loop here by using a itertools.product(sample inputs, variants) after the sample input and variant acquisition at the start of the test. Then we can create a helper function that executes this test body. The helper function takes a variant, a sample input, and whether the operation is inplace or not (so it knows whether to perform the copy. Then the test works like this:

- the variants are identified
- sample inputs are acquired
- if the op has an inplace variant, inplace sample inputs are acquired
- an itertools product of sample inputs x non-inplace variants invokes the helper function
- an itertools product of inplace sample inputs x inplace variants invokes the helper function

Does that make sense? Looking forward to hearing your thoughts.

You are correct that this will still redundantly create sample inputs that are common to both the out-of-place and in-place variant. Which is unfortunate from a performance standpoint, and your idea of marking SampleInputs as safe-for-inplace would have handled this problem better. However I think this is an OK penalty (for now, anyway).

Reply:
I'll try to do that. Will ping here if I stumble into another block.

Thanks!